### PR TITLE
[PYDRA] Statistics Volume Correction

### PR DESCRIPTION
**Pull request by MatthieuJoulot**
Pydra version of the statistics volume correction pipeline

### CODE REVIEW

File: clinica/pipelines/statistics_volume_correction/statistics_volume_correction_utils.py [hunk 1]
```diff
@@ -155,16 +190,17 @@ def produce_figures(nii_file, template, type_of_correction, t_thresh, c_thresh,
...
+def generate_output(t_map: str, figs: List[str], correction_name: str) -> None:
+    """Produce output
+
+    Parameters
+    ----------
+    t_map: str
+        Path to t-map on which whole pipeline was based
+    figs: list of str
+        Paths to figs to save
+    name: str
```
Comment (NicolasGensollen) [inline clinica/pipelines/statistics_volume_correction/statistics_volume_correction_utils.py:202]:
```suggestion
    correction_name: str
```

File: clinica/pydra/statistics_volume_correction/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+            c_thresh=parameters["FDRc"],
+        )
+    )
+    root = dirname(abspath(join(abspath(__file__), pardir, pardir)))
+    path_to_mask = join(root, "resources", "masks")
+    url_aramis = "https://aramislab.paris.inria.fr/files/data/img_t1_linear/"
+    FILE1 = RemoteFileStructure(
+        filename="mni_icbm152_t1_tal_nlin_sym_09a.nii.gz",
+        url=url_aramis,
+        checksum="3b244ee7e287319d36a25263744c468ef0ab2fe5a94b15a2138844db73b49adf",
```
Comment (NicolasGensollen) [inline clinica/pydra/statistics_volume_correction/pipeline.py:92]: Any chance this is what you're looking for, already implemented and ready to be plugged ?

https://github.com/aramis-lab/clinica/blob/a499f53eea49fd4151db30642239c34d7d3fa323/clinica/pydra/tasks.py#L20-L28

**Review by NicolasGensollen — COMMENTED** [review]
Thanks @MatthieuJoulot !
I think this is looking good. I made a super quick first pass with a few comments.
Will take a closer look next week !

File: clinica/pipelines/statistics_volume_correction/statistics_volume_correction_utils.py [hunk 3]
```diff
@@ -26,21 +35,31 @@ def peak_correction(t_map, t_threshold, output_name=None):
     else:
         filename = join("./peak_corrected_" + str(t_threshold) + basename(t_map))
     nib.save(new_data, filename)
-    return abspath(filename)
+    nii_file = abspath(filename)
```
[the remark this answers is not in the document]
Comment (MatthieuJoulot) [reply clinica/pipelines/statistics_volume_correction/statistics_volume_correction_utils.py:38]: I don't know but I guess I'll remove it

File: clinica/pydra/statistics_volume_correction/pipeline.py (hunk 2 above)
Comment (MatthieuJoulot) [reply clinica/pydra/statistics_volume_correction/pipeline.py:92]: I'll check it out

Comment (MatthieuJoulot) [reply clinica/pydra/statistics_volume_correction/pipeline.py:92]: I'm not sure it is what I want. I implemented this part the way it was made before, without any task. It does work that way. 
It feels like this is part of the logic I want. Maybe some hints about how to work this out are in `t1-linear` ? I'll check it out

Comment (MatthieuJoulot) [reply clinica/pydra/statistics_volume_correction/pipeline.py:92]: The file returned by this function is not the right one. Maybe we can discuss this tomorrow ?

Comment (NicolasGensollen) [reply clinica/pydra/statistics_volume_correction/pipeline.py:92]: You're right, it isn't the same file... 😓 The difference is minimal though (one letter actually...).

I think we should be consistent, so I would still implement this as a task function, either by passing a parameter to select the file, or by having two functions. Or something in between like this:

```python
@task
@annotate({"return": {"mni_template_file": PurePath}})
def download_mni_template_2009a() -> PurePath:
    return _download_mni_template(version="09a")

@task
@annotate({"return": {"mni_template_file": PurePath}})
def download_mni_template_2009c() -> PurePath:
    return _download_mni_template(version="09c")

def _download_mni_template(version: str) -> PurePath:
    return download_file(
        url=f"https://aramislab.paris.inria.fr/files/data/img_t1_linear/mni_icbm152_t1_tal_nlin_sym_{version}.nii.gz",
        to=str(Path.cwd() / f"mni_icbm152_t1_tal_nlin_sym_{version}.nii.gz"),
    )
```

File: clinica/pydra/statistics_volume_correction/pipeline.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    from clinica.utils.inputs import RemoteFileStructure, fetch_file
+    from clinica.utils.spm import spm_standalone_is_available, use_spm_standalone
+    from clinica.utils.stream import cprint
+
+    if spm_standalone_is_available():
+        use_spm_standalone()
+
+    # parameters = _check_pipeline_parameters(parameters)
+
+    input_name = "t_map"
```
Comment (NicolasGensollen) [inline clinica/pydra/statistics_volume_correction/pipeline.py:44]:
```suggestion
```

File: clinica/pydra/statistics_volume_correction/pipeline.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+    query = {"pattern": parameters["t_map"] + "*", "description": "statistics t map"}
+
+    input_spec = pydra.specs.SpecInfo(
+        name="Input",
+        fields=[
+            ("_graph_checksums", Any),
+            (input_name, dict, query, {"mandatory": True}),
```
Comment (NicolasGensollen) [inline clinica/pydra/statistics_volume_correction/pipeline.py:51]:
```suggestion
            ("t_map", dict, query, {"mandatory": True}),
```

File: clinica/pydra/statistics_volume_correction/pipeline.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+        utils.produce_figures_task(
+            name="produce_figure_FWE_peak_correction",
+            nii_file=wf.FWE_peak_correction_task.lzout.nii_file,
+            template=join(path_to_mask, FILE1.filename),
+            type_of_correction="FWE",
+            t_thresh=parameters["FWEp"],
+            c_thresh=np.nan,
+            n_cuts=parameters["n_cuts"],
+        )
+    )
```
Comment (NicolasGensollen) [inline clinica/pydra/statistics_volume_correction/pipeline.py:113]: Same, we can factorize here:

```python
for threshold in ("FWE", "FDR"):
    for kind in ("peak", "cluster"):
        t_thresh_key = f"{threshold}p" if kind == "peak" else "height_threshold"
        c_thresh = parameters[f"{threshold}c"] if kind == "cluster" else np.nan
        wf.add(
            utils.produce_figures_task(
                name=f"produce_figure_{threshold}_{kind}_correction",
                nii_file=getattr(wf, f"{threshold}_{kind}_correction_task").lzout.nii_file,
                template=join(path_to_mask, FILE1.filename),
                type_of_correction=threshold,
                t_thresh=parameters[t_thresh_key],
                c_thresh=c_thresh,
                n_cuts=parameters["n_cuts"],
            )
        )
        wf.add(
            utils.generate_output_task(
                 name=f"save_figure_{kind}_correction_{threshold}",
                 t_map=wf.lzin.t_map,
                 figs=getattr(wf, f"produce_figure_{threshold}_{kind}_correction").lzout.figs,
                 correction_name=f"{threshold}{kind[0]}",
             )
       )
```

File: clinica/pydra/statistics_volume_correction/statistics_volume_correction_cli.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    show_default=True,
+    help="Number of cuts along each direction",
+)
+@cli_param.option_group.common_pipelines_options
+@cli_param.option.working_directory
+@cli_param.option.n_procs
+def cli(
+    caps_directory: str,
+    t_map: str,
+    height_threshold: str,
```
Comment (NicolasGensollen) [inline clinica/pydra/statistics_volume_correction/statistics_volume_correction_cli.py:36]: I might be wrong, but it feels like this should be a float, doesn't it ?

**Review by NicolasGensollen — COMMENTED** [review]
Thanks @MatthieuJoulot !

File: clinica/pipelines/statistics_volume_correction/statistics_volume_correction_utils.py [hunk 8]
```diff
@@ -29,18 +39,29 @@ def peak_correction(t_map, t_threshold, output_name=None):
...
+
+    Parameters
+    ----------
+    t_map: str
+        Path to t-statistics nifti map
+    t_thresh: float
+        Threshold on t value
+    c_thresh: int
+        Minimal size of clusters after thresholding
+    output_name: str
```
Comment (NicolasGensollen) [inline clinica/pipelines/statistics_volume_correction/statistics_volume_correction_utils.py:58]:
```suggestion
    output_name: str, optional
```

File: clinica/pipelines/statistics_volume_correction/statistics_volume_correction_utils.py [hunk 9]
```diff
@@ -155,16 +191,17 @@ def produce_figures(nii_file, template, type_of_correction, t_thresh, c_thresh,
...
-        Produce output
-    Args:
-        t_map: (str) path to t-map on which whole pipeline was based
-        figs: (list of str) paths to figs to save
-        name: (str) name of the correction (ex: cluster_correction_FWE)
-
-    Returns:
-        Nothing
+def generate_output(t_map: str, figs: list, correction_name: str) -> None:
+    """Produce output
```
Comment (NicolasGensollen) [inline clinica/pipelines/statistics_volume_correction/statistics_volume_correction_utils.py:195]: I know it was like this before, but the docstring is even less informative than the function's name... 😱 
Maybe explain what is produced ?

File: clinica/pydra/statistics_volume_correction/task.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        `clinica.pipelines.statistics_volume_correction.statistics_volume_correction_utils.peak_correction`.
+    """
+    from clinica.pipelines.statistics_volume_correction.statistics_volume_correction_utils import (
+        peak_correction,
+    )
+
+    return peak_correction(t_map, t_threshold, output_name)
+
+
+@task
```
Comment (NicolasGensollen) [inline clinica/pydra/statistics_volume_correction/task.py:24]:
```suggestion
@task
@annotate({"return": {"nii_file": str}})
```

**Pull request merged**